rosbag2_test_common/include/rosbag2_test_common/subscription_manager.hpp

```diff
@@ -127,13 +127,56 @@ class SubscriptionManager
     return matched;
   }
 
-  std::future<void> spin_subscriptions()
+  /// @brief Wait until publishers will be connected to the subscriptions or timeout occur.
+  /// @param topic_names List of topic names
+  /// @param timeout Maximum time duration during which discovery should happen.
+  /// @param n_publishers_to_match Number of publishers each subscription should have for match.
+  /// @return true if subscriptions have specified number of publishers, otherwise false.
+  bool spin_and_wait_for_matched(
+    const std::vector<std::string> & topic_names,
+    std::chrono::duration<double> timeout = std::chrono::seconds(10),
+    size_t n_publishers_to_match = 1)
+  {
+    // Sanity check that we have valid input
+    if (topic_names.empty()) {
+      throw std::invalid_argument("List of topic names is empty");
+    }
+    for (const auto & topic_name : topic_names) {
+      if (subscriptions_.find(topic_name) == subscriptions_.end()) {
+        throw std::invalid_argument(
+                "Publisher's topic name = `" + topic_name + "` not found in expected topics list");
+      }
+    }
+
+    using clock = std::chrono::steady_clock;
+    auto start = clock::now();
+
+    rclcpp::executors::SingleThreadedExecutor exec;
+    bool matched = false;
+    while (!matched && ((clock::now() - start) < timeout)) {
+      exec.spin_node_some(subscriber_node_);
+
+      matched = true;
+      for (const auto & topic_name : topic_names) {
+        if (subscriptions_.find(topic_name) == subscriptions_.end() ||
+          subscriptions_[topic_name]->get_publisher_count() < n_publishers_to_match)
+        {
+          matched = false;
+          break;
+        }
+      }
```

Comment:
Consider the case where `spin_node_some` has no work to do; in that case, it may[1] quickly return, we will quickly run through the loop above, and then we'll go through the outer loop again quickly, burning a lot of CPU.  It is probably worthwhile to add a short sleep here (maybe a millisecond).

[1] Today that isn't too much of an issue, but with some of the executor speedup work this becomes more of a potential problem.

Reply:
@clalancette I think it would be better to replace it with `exec.spin_node_once(subscriber_node_, timeout);` instead of adding delay after `spin_node_some`. 
What do you think?

Reply:
@clalancette I have to admit that my proposal of replacing `spin_node_some` with `spin_node_once` was a bad idea.
It turns out that `spin_node_once` will block up until timeout in the test scenario and test execution time is increasing up to the timeout of 10 seconds for `wait_for_matched();`
I've implemented a fix as you suggested with small sleep for a few milliseconds to avoid busy loop.